In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
df_prods = pd.read_csv(r'/Users/emilyseitz/Instacart Basket Analysis /Data/Original Data/products.csv', index_col = False)

FileNotFoundError: [Errno 2] No such file or directory: '/Users/emilyseitz/Instacart Basket Analysis /Data/Original Data/products.csv'

In [3]:
df_ords = pd.read_csv(r'/Users/emilyseitz/Instacart Basket Analysis /Data/Prepared Data/orders_wrangled.csv', index_col = False)

FileNotFoundError: [Errno 2] No such file or directory: '/Users/emilyseitz/Instacart Basket Analysis /Data/Prepared Data/orders_wrangled.csv'

In [116]:
path = r'/Users/emilyseitz/Instacart Basket Analysis '

In [114]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [21]:
# Create a dataframe
df_test = pd.DataFrame()

In [23]:
# Create a mixed type column
df_test['mix'] = ['a','b', 1, True]


In [24]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [26]:
# check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [29]:
df_test['mix'] = df_test['mix'].astype('str')

In [30]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [106]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [32]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [33]:
df_prods.median()

/var/folders/_p/b1q7n89x4hl72qqvl_1wc7lm0000gn/T/ipykernel_28399/3387230063.py:1: FutureWarning: The default value of numeric_only in DataFrame.median is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df_prods.median()


product_id       24845.0
aisle_id            69.0
department_id       13.0
prices               7.1
dtype: float64

In [107]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [36]:
df_prods.shape

(49693, 5)

In [109]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [38]:
df_prods_clean.shape

(49677, 5)

In [108]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [40]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [45]:
df_prods_clean.shape

(49677, 5)

In [110]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [111]:
df_prods_clean_no_dups.shape

(49672, 5)

In [118]:
#Export dataframe 
df_prods.to_csv(os.path.join(path, 'Data','Prepared Data','product_cleaned.csv'))

In [122]:
# Excercise 4.5 -Question 2
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

## The days_since_prior_order has a lower number 1.5 for the 75% percentile which is lower than the 25% and 50% percentile markers- it also throws off the max to be lower than the 25% marker.

## The order_hour_of_day looks a little strange in the distribution doesn't exceend 2:30 and the distribution matches as well.

## The order_id has a much higher number in the 25% pecentile and a high std.

In [125]:
# check for mixed data types - Question 3
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0           0   2539329        1    prior             1          2   
1           1   2398795        1    prior             2          3   
2           2    473747        1    prior             3          3   
3           3   2254736        1    prior             4          4   
4           4    431534        1    prior             5          4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [126]:
# check for mixed data types - Question 3
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

## no code was executed so there are not mixed examples in the columns.

## Question 4- based on the examples provided on how to check for mixed data types It looks not applicable. However, The days_since_prior_order with NaN looks like it might need to be converted to an integer. 


In [127]:
# Question 5- Missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [129]:
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [130]:
df_nan

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0                 0   2539329        1    prior             1          2   
11               11   2168274        2    prior             1          2   
26               26   1374495        3    prior             1          1   
39               39   3343014        4    prior             1          6   
45               45   2717275        5    prior             1          3   
...             ...       ...      ...      ...           ...        ...   
3420930     3420930    969311   206205    prior             1          4   
3420934     3420934   3189322   206206    prior             1          3   
3421002     3421002   2166133   206207    prior             1          6   
3421019     3421019   2227043   206208    prior             1          1   
3421069     3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 8 columns]

## days_since_prior_order has 206,209 occurrences of missing entries. This could be due to it's an inital first time order and there is no history of an earlier order.

In [131]:
# Question 5 -fix missing values
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [143]:
# Question 6- part address missing values
df_ords.dropna(inplace = True)

## I replaced all missing values from all columns, with the True option to overwrite the original data frame.

In [165]:
df_ords.shape

(3214874, 8)

In [183]:
#Question 7 on duplicates
df_ords_clean_no_dups = df_ords_clean.drop_duplicates()

In [180]:
df_ords_clean_no_dups.shape

(3214874, 8)

## the # of rows did not change since the missing values in cells were overwritten but no rows were deleted.

In [185]:
#Question 8- on duplicates
df_dups = df_ords_clean[df_ords_clean.duplicated()]

In [182]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

## There were no duplicates found in the cleaned data frame by the above code.

In [186]:
#Export dataframe for ords
df_ords.to_csv(os.path.join(path, 'Data','Prepared Data','orders_cleaned.csv'))